src/sentry/api/endpoints/organization_teams.py

```diff
@@ -142,56 +145,92 @@ def post(self, request: Request, organization, **kwargs) -> Response:
+        :qparam bool set_team_admin: If this is true, the user is added to the as a Team Admin
+                                instead of regular member
         :auth: required
         """
         serializer = TeamPostSerializer(data=request.data)
 
-        if serializer.is_valid():
-            result = serializer.validated_data
+        if not serializer.is_valid():
+            raise ValidationError(serializer.errors)
 
-            try:
-                with transaction.atomic():
-                    team = Team.objects.create(
-                        name=result.get("name") or result["slug"],
-                        slug=result.get("slug"),
-                        idp_provisioned=result.get("idp_provisioned", False),
-                        organization=organization,
-                    )
-            except IntegrityError:
-                return Response(
-                    {
-                        "non_field_errors": [CONFLICTING_SLUG_ERROR],
-                        "detail": CONFLICTING_SLUG_ERROR,
-                    },
-                    status=409,
+        result = serializer.validated_data
+        set_team_admin = result.get("set_team_admin")
+
+        if set_team_admin:
+            if not features.has("organizations:team-roles", organization) or not features.has(
+                "organizations:team-project-creation-all", organization
+            ):
+                raise ResourceDoesNotExist(
+                    detail="You do not have permission to join a new team as a team admin"
+                )
+            if not self.should_add_creator_to_team(request):
+                raise ValidationError(
+                    {"detail": "You do not have permission to join a new team as a Team Admin"},
                 )
-            else:
-                team_created.send_robust(
-                    organization=organization, user=request.user, team=team, sender=self.__class__
+        try:
+            with transaction.atomic():
+                team = Team.objects.create(
+                    name=result.get("name") or result["slug"],
+                    slug=result.get("slug"),
+                    idp_provisioned=result.get("idp_provisioned", False),
+                    organization=organization,
                 )
-            if self.should_add_creator_to_team(request):
-                try:
+        except (IntegrityError, MaxSnowflakeRetryError):
+            raise ConflictError(
+                {
+                    "non_field_errors": [CONFLICTING_SLUG_ERROR],
+                    "detail": CONFLICTING_SLUG_ERROR,
+                }
+            )
+        else:
+            team_created.send_robust(
+                organization=organization,
+                user=request.user,
+                team=team,
+                sender=self.__class__,
+            )
+
+        if self.should_add_creator_to_team(request):
+            cleanup_needed = False
```

Comment:
nit: It's a little cleaner to default `cleanup_needed` to `True` and only mark it `False` at the end of the `try` block. That way you won't need to generic Exception block.

Reply:
```
cleanup = False
try:
    # code
except Exc1:
    # code
    cleanup = True
except Exception:
    cleanup = True
finally:
    if cleanup:
        # cleanup
```
vs
```
cleanup = True
try:
    # code
    cleanup = False
except Exc1 : 
    # code
finally:
    if cleanup:
        # cleanup
```

Reply:
ok, I changed it to the bottom. The only thing is the endpoint should only delete the team when a generic error occurs if `set_team_admin` is true. The flow is like:
```
cleanup = True
try:
    # code
    cleanup = False
except Exc1 : 
    # code
finally:
    if set_team_admin and cleanup:
        # delete team
```